src/main/java/com/jcraft/jsch/Channel.java

```diff
@@ -550,18 +545,11 @@ public boolean isClosed(){
   static void disconnect(Session session){
     Channel[] channels=null;
     int count=0;
-    synchronized(pool){
-      channels=new Channel[pool.size()];
-      for(int i=0; i<pool.size(); i++){
-        try{
-          Channel c=pool.elementAt(i);
-          if(c.session==session){
-            channels[count++]=c;
-          }
-        }
-        catch(Exception e){
-        }
-      } 
+    channels=new Channel[pool.size()];
```

Comment:
This looks problematic, in that the size of `pool` could increase after the call to `pool.size()` here and before the iteration over `pool.values()` below, potentially leading to an `ArrayIndexOutOfBoundsException` being thrown.

I think that instead of accumulating all the channels with a matching session into an array, and then iterating over this array to call `disconnect()`, we could simply hoist the call to `disconnect()` directly in the first iteration (that checks for matching sessions).

I believe the only reason it was done this way (storing matching sessions in an array and then iterating over said array) was so that the calls to `disconnect()` didn't happen while `pool` was locked.

Since with this change you are no longer locking the `pool` object, it should be safe to directly call `disconnect()` while iterating through `pool`.

Reply:
I got it, I will improve it. Please delete my misscreated PullRequest.